### PR TITLE
Update the usage of the transform dialect interpreter

### PR DESCRIPTION
**Pull request by nicolasvasilache**
Upstream is shifting to using the simpler form of the transform dialect interpreter introduced in https://github.com/llvm/llvm-project/pull/68661

Update IREE's codegen usage of the interpreter as well as tests which now require a named sequence entry point.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectInterpreterPass.cpp [hunk 1]
```diff
@@ -20,62 +21,54 @@ namespace {
...
+    auto dialect = context->getOrLoadDialect<
+        mlir::iree_compiler::IREE::Codegen::IREECodegenDialect>();
+    auto maybeTransformLibrary =
+        dialect->getOrLoadTransformLibraryModule(libraryFileName);
+
+    Operation *payloadRoot = getOperation();
+    ModuleOp transformModule =
+        succeeded(maybeTransformLibrary) ? *maybeTransformLibrary : ModuleOp();
+    Operation *transformEntryPoint = transform::detail::findTransformEntryPoint(
+        getOperation(), transformModule, entryPoint);
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectInterpreterPass.cpp:57]: nit: I would add an assert here for successfully finding the entry point. Helps for when users try to invoke this pass in isolation.

File: compiler/src/iree/compiler/Codegen/Common/MaterializeUserConfigs.cpp [hunk 2]
```diff
@@ -72,36 +58,44 @@ struct MaterializeUserConfigsPass
         getAllEntryPoints(moduleOp);
     MLIRContext *context = moduleOp.getContext();
 
+    LDBG("MaterializeUserConfigsPass on variant: " << variantOp);
+    // LLVM_DEBUG(llvm::interleaveComma(exportOps, DBGS() << "exportOps: "));
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/MaterializeUserConfigs.cpp]: nit: remove commented line

File: compiler/src/iree/compiler/Codegen/Common/Passes.td [hunk 3]
```diff
@@ -445,35 +445,17 @@ def TransformDialectInterpreter :
...
-            "allows user control on what transformation to apply in debug "
-            "mode, without requiring intimate knowledge of the IREE nested "
-            "pass pipeline.\\n"
-            "If empty (normal operation mode), select the container of the "
-            "top-level transform op.">
+    Option<"entryPoint", "entry-point", "std::string",
+           /*default=*/[{"__transform_main"}],
+           "Entry point of the pass pipeline.">,
+    Option<"libraryFileName", "library-file-name", "std::string",
+           /*default=*/[{""}], "">,
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/Passes.td]: nit: Add a description for the pass option, like
```
File path to load a library of transform dialect strategies from.
```

**Review by qedawkins — COMMENTED** [review]
Holding off on approval until all of the CI failures are unwound, but the material changes in this PR (to the interpreter/user config passes) look good to me.

Additionally I expect you'll need to remove/disable the sample here for now: https://github.com/openxla/iree/tree/main/samples/transform_dialect

File: tests/transform_dialect/cuda/softmax_partial.mlir [hunk 4]
```diff
@@ -14,7 +15,8 @@
 /// Constant JIT'ing must be disabled because the transform-dialect debug
 /// flags leak to the JIT session, which doesn't know what to do with them.
 // RUN:     --iree-codegen-llvmgpu-enable-transform-dialect-jit=false \
-// RUN:     --iree-codegen-use-transform-dialect-strategy=%p/softmax_partial_codegen_spec.mlir | \
+// RUN:     --iree-codegen-transform-dialect-library=%p/softmax_partial_codegen_spec.mlir \
+// RUN:     --iree-codegen-use-transform-dialect-strategy=codegen \
```
Comment (qedawkins) [inline tests/transform_dialect/cuda/softmax_partial.mlir]: I think your CI failures are from missing a pipe `|` here

**Review by qedawkins — APPROVED** [review]
Approving, contingent on removing the LLVM cherry pick.

File: compiler/src/iree/compiler/Codegen/Common/GPU/test/BUILD.bazel [hunk 5]
```diff
@@ -6,8 +6,8 @@
 
 # Tests for common transforms.
 
-load("//build_tools/bazel:iree_lit_test.bzl", "iree_lit_test_suite")
 load("//build_tools/bazel:enforce_glob.bzl", "enforce_glob")
+load("//build_tools/bazel:iree_lit_test.bzl", "iree_lit_test_suite")
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/test/BUILD.bazel:10]: One more question, what is the need for all of these changes to the bazel build files?

**Comment by qedawkins** [comment]
The most recent integrate should have pulled in the cherry-picked changes: https://github.com/openxla/iree/commit/18b3dd78832c4f6f4c8961901d328bca96c7efc3

File: compiler/src/iree/compiler/Codegen/Common/GPU/test/BUILD.bazel (hunk 5 above)
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/GPU/test/BUILD.bazel:10]: I have had this local script for a while now that I always run before committing:

``` 
iree-presubmits() {
  python ${IREE_SOURCE_DIR}/build_tools/bazel_to_cmake/bazel_to_cmake.py
  (cd ${IREE_SOURCE_DIR} && ./build_tools/scripts/run_buildifier.sh)
}
```

It picked up those changes.

**Pull request merged**